src/libsyntax/parse/parser.rs

```diff
@@ -4173,6 +4174,29 @@ impl<'a> Parser<'a> {
                     }));
                     self.bump();
                 }
+                token::Interpolated(ref nt) => {
+                    match **nt {
+                        token::Nonterminal::NtPath(ref nt_path) => {
+                            let trait_ref = ast::TraitRef {
+                                path: nt_path.clone(),
+                                ref_id: ast::DUMMY_NODE_ID,
+                            };
+                            let poly_trait_ref = ast::PolyTraitRef {
+                                bound_lifetimes: Vec::new(),
+                                trait_ref: trait_ref,
+                                span: mk_sp(self.span.lo, self.span.hi),
+                            };
+                            let modifier = if ate_question {
+                                TraitBoundModifier::Maybe
+                            } else {
+                                TraitBoundModifier::None
+                            };
+                            result.push(TraitTyParamBound(poly_trait_ref, modifier));
+                            should_bump = true;
+                        }
+                        _ => break,
+                    }
+                }
                 token::ModSep | token::Ident(..) => {
```

Comment:
Instead of adding the `token::Interpolated(ref nt) => { ... }` arm, I believe you could just change this pattern to
```rust
        _ if self.token.is_path_start() || self.token.is_keyword(keywords::For) => { ... }
```